lib/web/apiserver.go

```diff
@@ -2688,6 +2692,60 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
 	})
 }
 
+// ProvisionTokenHandler is a authenticated handler that is called for some existing Token
+type ProvisionTokenAuthedHandler func(w http.ResponseWriter, r *http.Request, p httprouter.Params, site reversetunnel.RemoteSite) (interface{}, error)
+
+// WithProvisionTokenAuth ensures that request is authenticated with a provision token.
+// Provision tokens, when used like this are invalidated as soon as used.
+// Doesn't matter if the underlying response was a success or an error.
+func (h *Handler) WithProvisionTokenAuth(fn ProvisionTokenAuthedHandler) httprouter.Handle {
+	return httplib.MakeHandler(func(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+		ctx := r.Context()
+		logger := h.log.WithField("request", fmt.Sprintf("%v %v", r.Method, r.URL.Path))
+
+		creds, err := roundtrip.ParseAuthHeaders(r)
+		if err != nil {
+			logger.WithError(err).Warn("No auth headers.")
+			return nil, trace.AccessDenied("need auth")
+		}
+
+		if err := h.consumeTokenForAPICall(ctx, creds.Password); err != nil {
+			h.log.WithError(err).Warn("Failed to authenticate.")
+			return nil, trace.AccessDenied("need auth")
+		}
+
+		clusterName := p.ByName("site")
+		if clusterName == currentSiteShortcut {
+			res, err := h.GetProxyClient().GetClusterName()
+			if err != nil {
+				h.log.WithError(err).Warn("Failed to query cluster name.")
+				return nil, trace.Wrap(err)
+			}
+			clusterName = res.GetClusterName()
+		}
+
+		site, err := h.cfg.Proxy.GetSite(clusterName)
+		if err != nil {
+			h.log.WithError(err).WithField("cluster-name", clusterName).Warn("Failed to query site.")
+			return nil, trace.Wrap(err)
+		}
+
+		return fn(w, r, p, site)
+	})
+}
+
+func (h *Handler) consumeTokenForAPICall(ctx context.Context, tokenName string) error {
```

Comment:
Get+Delete aren't atomic, would this result into potential token reuse if multiple clients reach here concurrently?

Reply:
Yeah, that's possible.

This method reports no error if `GetToken and DeleteToken` return no error.
DeleteToken calls Delete User Token and Delete (Provision) Token

For Delete (Provision) Token, it has a check for a Not Found error
However, having a NotFound error depends on the backend Implementation
From our current implementations, only Firestore does not guarantee such error when trying to delete an already deleted document.
To fix this: https://github.com/gravitational/teleport/pull/14932

After merging the above PR, we can be confident that multiple clients won't be able to use the same token.


Regarding User Tokens, I think we should remove them as a viable option.
Let's discuss that in the thread below